### PR TITLE
Alternative: Add imaginary literal syntax

### LINKED ISSUE #941: imaginary number literals
R allows users to write a complex number as `1.2 + 3.7i`, where `3.7i` is understood to be an imaginary literal.  Python and MATLAB use `j` in place of `i`.  The Stan language currently requires this to be written as `to_complex(1.2, 3.7)`.  

Stan could support imaginary literals of the form `<fp>i` where `<fp>` is a floating-point literal.  Examples include`1i`, `3.7i`, and `2.9876e-12i`.  The imaginary unit must be written `1i` rather than just `i`.    That is, the postfix `i` acts much like the prefix `0x` marks the literal `238` as hexadecimal in the C++ literal `0x238` or the postfix `L` in `237L` marks `237` as a long integer in C++.

The whole expression `1.2 + 3.7i` would consist of a real literal and imaginary literal, which when added together, produce the same value as `std::complex(1.2, 3.7)`.   The number 3.7 would be instantiated as the complex number `std::complex(0, 3.7)` with real part 0 and imaginary part 3.7.    Thus `1.2 + 3.7i` would compile to `1.2 + std::complex(0, 3.7)`.   This has the same behavior as promoting 1.2 to complex, i.e., `std::complex(1.2, 0) + std::complex(0, 3.7)`, but can be accomplished without explicitly promoting 1.2 to a complex number because `std::operator+()` has appropriate overloads for mixed real and complex arithmetic.  

The complex constructor `std::complex()` is `constexpr` and so is `std::operator+(U, V)` where either of the `U` or `V` is complex.    That means the C++ compiler should be able to evaluate the complete complex literal `1.2 + 3.7i` to a complex object at compile time.

### PR DESCRIPTION
**Pull request by WardBrian**
This is an alternative implementation to #959 which I think I prefer more now that I look at it.

This is a follow on to #940 and closes #941. It adds the syntax `(realconstant | intconstant) i` for representing imaginary numbers. Examples:

```
  complex zi = 1 + 3.14i;
  zi = zi * 0i;
  real x = get_real(3i - 40e-3i);
```

The implementation generates a call to `to_complex` with the first argument set to zero during the codegen phase. This will be constexpr'd away by the C++ compiler. For most of the compiler, this is treated by mirroring the way RealNumeral is in terms of type/semantic checking. 

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [x] If a user-facing facing change was made, the documentation PR is here: 
    This will follow on with https://github.com/stan-dev/docs/pull/378

## Release notes

Added the ability to write imaginary numbers as literals, such as "3.1i"

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/stan_math_backend/Expression_gen.ml [hunk 1]
```diff
@@ -540,6 +540,7 @@ and pp_expr ppf Expr.Fixed.({pattern; meta} as e) =
   match pattern with
   | Var s -> pf ppf "%s" s
   | Lit (Str, s) -> pf ppf "%S" s
+  | Lit (Imaginary, s) -> pf ppf "to_complex(0, %s)" s
   | Lit (_, s) -> pf ppf "%s" s
```
Comment (nhuurre) [inline src/stan_math_backend/Expression_gen.ml:544]: [optional] I prefer avoiding wildcards when there aren't too many variants.
```suggestion
  | Lit ((Int | Real), s) -> pf ppf "%s" s
```

**Review by nhuurre — APPROVED** [review]
Looks good, one question.
What's the status of complex numbers in the TFP backend? The complex scalar PR didn't add any TFP tests but if `to_complex()` works then this PR should do something there too. Python has imaginary literals with `j` suffix but the backend does a type cast for floats so I'm not sure what exactly is the correct output.

**Comment by WardBrian** [comment]
Complex numbers currently fail in the TFP backend because their functions weren't added to the mapping of stanlib functions:

https://github.com/stan-dev/stanc3/blob/a1740dbd9929f86106948300144320512e5dc94c/src/tfp_backend/Transform_mir.ml#L35

Though this mapping is also missing most basic math functions, like abs, exp, etc, so even adding simple translations for to_complex, get_real, and get_imag still doesn't allow the complex_scalar.stan test to work on that backend. I think that is worth a separate PR later

**Comment by WardBrian** [comment]
@nhuurre - just to confirm is this good to merge?

**Comment by nhuurre** [comment]
Yes. No reason to worry about imaginary literals in TFP if the complex numbers are unsupported anyway.

**Pull request merged**